### PR TITLE
Replace for loops with iterators

### PR DESCRIPTION
**Pull request by federicomenaquintero**
The first ones are straightforward.  The last one is a bit more involved; feel free to ignore it if you think it makes the code harder to read (the tests may still be useful).

I wanted to use `.windows()` on the `data` slice but couldn't find a nice way to limit to the `range_length`.

### CODE REVIEW

**Comment by federicomenaquintero** [comment]
(Re-pushed to fix the test for looks_like_text(), to ensure that passing an empty data still returns true.)

**Comment by federicomenaquintero** [comment]
Found a way to use windows :)

**Comment by ebassi** [comment]
"Much functional, such iterators, wow"

Thanks, this looks a lot less "let's rewrite C in Rust". :-)

File: src/magic.rs [hunk 1]
```diff
@@ -41,43 +41,39 @@ struct MagicRule {
 
 impl MagicRule {
     fn matches_data(&self, data: &[u8]) -> bool {
+        // Do we need the value_length at all, if it's implicit in
+        // value.len() and checked by the parsers?
```
Comment (ebassi) [inline src/magic.rs:45]: That's a fair question, and I think the answer is: no, we don't need to store the length at all.

File: src/magic.rs [hunk 2]
```diff
@@ -41,43 +41,39 @@ struct MagicRule {
...
-                        let masked_value = self.value[j] & m[j];
-                        let masked_data = data[j + i] & m[j];
-                        if masked_value != masked_data {
-                            res = false;
-                            break;
+                    if looking_at
+                        .iter()
+                        .zip(m.iter())
+                        .map(|(l, m)| *l & *m)
+                        .eq(self.value.iter().map(|v| *v)) {
```
Comment (ebassi) [inline src/magic.rs]: This is kind of black magic and a bit hard to read for me.

The logic is that both the value in the rule and the data we're matching must be masked, but the `map()` seems to be masking only the `looking_at` vector.

File: src/magic.rs (hunk 1 above)
Comment (federicomenaquintero) [reply src/magic.rs]: I've removed the comment but kept the assertions.  Per the commit message,

```
Remove comment that wonders whether checking value/mask sizes is necessary

It is.  Right now it is done in the parser, but hopefully we can move
to model where a) the parser enforces that; b) a MagicRule::new() or
MagicRuleBuilder asserts that.
```

File: src/magic.rs (hunk 2 above)
Comment (federicomenaquintero) [reply src/magic.rs]: Oh, I see, indeed the original code did what you say :)  I've updated it with a more legible version.

Comment (federicomenaquintero) [reply src/magic.rs]: Also I updated the test for masks to actually need masking both the value and the data.

**Comment by federicomenaquintero** [comment]
Another update - I misunderstood range_length.  Added a test for it and fixed the implementation.

**Pull request merged**